### PR TITLE
Improve sections formatting inside keywords

### LINKED ISSUE #41: Formatter: multiple issues in `for` loop
1. Extra blank line is not deleted.
2. Wrong indent for several subsequent `for` loops.
Selection_001

https://apps.axibase.com/chartlab/42a161bc/12/

### PR DESCRIPTION
**Pull request by PolinaShneider**
Closes #41 

## Update:

Simplified formatter's logic — now consequent blank lines are deleted at the end after processing entire config.

## What was done:

Changed formatting behaviour of sections inside keywords structures such as `csv|endcsv`, `for|endfor`, `list|endlist` etc

Now blank line before section declaration isn't inserted immediately after start of keyword structure. 

Blank lines separators of subsequent sections are preserved.

image

### CODE REVIEW

**Comment by annstriganova** [comment]
It seems that [items](https://github.com/axibase/charts-language-service/issues/41#issuecomment-528329277) **3** and **5** are not implemented. Is it too difficult?

**Comment by PolinaShneider** [comment]
@annstriganova just to make sure, I correctly understand you about 3 and 5 formatting features. 

- 3 is current behaviour for all sections. For control structures in `[configuration]`, `[group]` and `[widget]` it should remain the same

-  For control structures  in sections, having depth more than `[widget]` I need to implement feature 5 — aligning `for|if|csv` etc with contents to parent section.

**Comment by annstriganova** [comment]
> - 3 is current behaviour for all sections. For control structures in `[configuration]`, `[group]` and `[widget]` it should remain the same

Yes, no changes in this case.

> -  For control structures  in sections, having depth more than `[widget]` I need to implement feature 5 — aligning `for|if|csv` etc with contents to parent section.

I'd suggest to implement feature 5 only for `for`, `list`, `var`, `csv`, since I don't know real use cases, when this constructions are used in `[series]`, `[column]` and etc.

**Comment by annstriganova** [comment]
Could it be used the following algorithm:
1. if we at first line inside keyword, do not add blank line;
2. if we at line with keyword end, remove previous blank line
64519216-659eac80-d2fc-11e9-8b6a-186be6145a64
?

**Comment by PolinaShneider** [comment]
Yes, that's possible. Will implement it

**Comment by PolinaShneider** [comment]
And what about nested control structures inside sections width depth > `[widget]`? 

Not sure, how to handle formatting feature №5 in this case: 

image

**Comment by annstriganova** [comment]
`if` is inside `[column]`, so align as column setting.

**Comment by PolinaShneider** [comment]
> I'd suggest to implement feature 5 only for `for`, `list`, `var`, `csv`

Which of these control structures can potentially be nested? I suppose, only `for`. Thinking of a workaround for this case (`for` in `[column]` is aligned to parent `[widget]`)

image

**Comment by PolinaShneider** [comment]
Formatting feature №5 is implemented and ready to be tested

**Comment by PolinaShneider** [comment]
Reverted feature 5, because with the existing formatting logic handling nested and sibling structures becomes a nightmare. Also it's unclear how to determine at what moment structure outdent should be terminated.

I think, it can be implemented in a separate PR. Probably after designing a syntax tree.

The rest of the features are implemented. @annstriganova could you review it, please?

File: src/formatter.ts [hunk 1]
```diff
@@ -182,6 +203,38 @@ export class Formatter {
         return this.formattedText.join("\n");
     }
 
+    /**
+     * Inserts blank line after keyword end if needed
+     */
+    private formatControlStructure(): void {
```
Comment (annstriganova) [inline src/formatter.ts]: `formatControlStructure` is confusing, `insertBlankLineAfterKeywordEnd` is more suitable.

File: src/formatter.ts [hunk 2]
```diff
@@ -488,11 +559,18 @@ export class Formatter {
         this.lines.splice(this.currentLine + 1, 0, line);
     }
 
+    /**
+     * Determines whether blank line should be inserted before setting inside control structure
+     */
+    private get noBlankLineBeforeSectionInKeyword(): boolean {
+        return this.insideKeyword && this.keywordSections.length === 1 && !this.keywordSections[0].hasSettingBefore;
```
Comment (annstriganova) [inline src/formatter.ts]: Is it necessary check?

```ts
this.keywordSections.length === 1
```

Comment (PolinaShneider) [reply src/formatter.ts]: We may have different formatting behaviour (whether to insert blank line before or not) only for first section inside keyword. Second and next sections should be separated with blank lines in any case.

Comment (annstriganova) [reply src/formatter.ts]: Please, add a test, covering this case.

Comment (PolinaShneider) [reply src/formatter.ts]: Added test cases. Without this check `"Multiple sections inside keyword, only first without blank line"` fails as intended

**Comment by annstriganova** [comment]
@PolinaShneider, create an issue describing inevitability of syntax tree, so we can discuss it as a team.

**Comment by PolinaShneider** [comment]
Created [issue](https://github.com/axibase/charts-language-service/issues/57) for discussion

**Comment by rodionos** [comment]
@PolinaShneider - pls fix the merge conflict

**Pull request merged**